iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1097 @@
-
-                await s_eventLoopGroup.Value.RegisterAsync(clientWebSocketChannel).ConfigureAwait(true);
+            Dictionary<string, string> mergedProperties = systemProperties.Concat(message.Properties)
+                .ToLookup(x => x.Key, x => x.Value)
+                .ToDictionary(x => x.Key, g => g.First());
 
-                return clientWebSocketChannel;
-            };
-        }
-
-        private void ScheduleCleanup(Func<Task> cleanupTask)
-        {
-            Func<Task> currentCleanupFunc = _cleanupFunc;
-            _cleanupFunc = async () =>
-            {
-                await cleanupTask().ConfigureAwait(true);
+            string properties = UrlEncodedDictionarySerializer.Serialize(mergedProperties);
 
-                if (currentCleanupFunc != null)
-                {
-                    await currentCleanupFunc().ConfigureAwait(true);
-                }
-            };
-        }
+            string msg = properties.Length != 0
+                ? topicName.EndsWith("/", StringComparison.Ordinal) ? $"{topicName}{properties}/" : $"{topicName}/{properties}"
+                : topicName;
 
-        private async Task CleanUpAsync()
-        {
-            try
+            if (Encoding.UTF8.GetByteCount(msg) > MaxTopicNameLength)
             {
-                await _closeRetryPolicy.RunWithRetryAsync(CleanUpImplAsync).ConfigureAwait(true);
+                throw new IotHubClientException($"TopicName for MQTT packet cannot be larger than {MaxTopicNameLength} bytes, " +
+                    $"current length is {Encoding.UTF8.GetByteCount(msg)}." +
+                    $" The probable cause is the list of message.Properties and/or message.systemProperties is too long.", false, IotHubStatusCode.MessageTooLarge);
             }
-            catch (Exception ex) when (!Fx.IsFatal(ex))
-            {
-            }
-        }
 
-        private Task CleanUpImplAsync()
-        {
-            return _cleanupFunc == null
-                ? Task.CompletedTask
-                : _cleanupFunc();
+            return msg;
         }
 
-        private bool TryStateTransition(TransportState fromState, TransportState toState)
+        private bool ParseResponseTopic(string topicName, out string rid, out int status, out int version)
         {
-            return (TransportState)Interlocked.CompareExchange(ref _state, (int)toState, (int)fromState) == fromState;
-        }
+            rid = "";
+            status = 500;
+            version = 0;
 
-        private void EnsureValidState(bool throwIfNotOpen = true)
-        {
-            if (State == TransportState.Error)
+            Match match = _twinResponseTopicRegex.Match(topicName);
+            if (match.Success)
             {
-                _fatalException.Throw();
-            }
+                status = Convert.ToInt32(match.Groups[1].Value, CultureInfo.InvariantCulture);
```

Comment:
Magic numbers at least require some explanation, if not an alias variable to describe what it is.